constraint-solver/src/solver.rs

```diff
@@ -96,8 +110,187 @@ pub enum Error {
+}
+
+struct BooleanExtractedSolver<T, V, S> {
+    solver: S,
+    boolean_var_dispenser: BooleanVarDispenser,
+    _phantom: std::marker::PhantomData<(T, V)>,
+}
+
+impl<T, V, S> BooleanExtractedSolver<T, V, S>
+where
+    T: RuntimeConstant + VarTransformable<V, Variable<V>>,
+    T::Transformed: RuntimeConstant<FieldType = T::FieldType>,
+    V: Clone + Eq,
+    S: Solver<T::Transformed, Variable<V>>,
+{
+    fn new(solver: S) -> Self {
+        Self {
+            solver,
+            boolean_var_dispenser: BooleanVarDispenser::new(),
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<T, V, S> RangeConstraintProvider<T::FieldType, V> for BooleanExtractedSolver<T, V, S>
+where
+    T: RuntimeConstant,
+    S: RangeConstraintProvider<T::FieldType, Variable<V>>,
+    V: Clone,
+{
+    fn get(&self, var: &V) -> RangeConstraint<T::FieldType> {
+        self.solver.get(&Variable::from(var.clone()))
+    }
+}
+
+impl<T, V, S: Display> Display for BooleanExtractedSolver<T, V, S> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Boolean extracted solver:\n{}", self.solver)
+    }
+}
+
+impl<T, V, S> Solver<T, V> for BooleanExtractedSolver<T, V, S>
+where
+    T: RuntimeConstant + VarTransformable<V, Variable<V>> + Display,
+    T::Transformed: RuntimeConstant<FieldType = T::FieldType>
+        + VarTransformable<Variable<V>, V, Transformed = T>
+        + Display,
+    V: Ord + Clone + Eq + Hash + Display,
+    S: Solver<T::Transformed, Variable<V>>,
+{
+    /// Solves the system and ignores all assignments that contain a boolean variable
+    /// (either on the LHS or the RHS).
+    fn solve(&mut self) -> Result<Vec<VariableAssignment<T, V>>, Error> {
+        let assignments = self.solver.solve()?;
+        Ok(assignments
+            .into_iter()
+            .filter_map(|(v, expr)| {
+                let v = v.try_to_original()?;
+                let expr = expr.try_transform_var_type(&mut |v| v.try_to_original())?;
+                Some((v, expr))
+            })
+            .collect())
+    }
+
+    fn add_algebraic_constraints(
+        &mut self,
+        constraints: impl IntoIterator<Item = GroupedExpression<T, V>>,
+    ) {
+        let constraints = constraints
+            .into_iter()
+            .flat_map(|constr| {
+                let constr = constr.transform_var_type(&mut |v| v.clone().into());
+                let extracted =
+                    try_extract_boolean(&constr, &mut || self.boolean_var_dispenser.next_var())
+                        .inspect(|_| {
+                            // Make sure to range-constrain the variable
+                            self.solver.add_range_constraint(
+                                &self.boolean_var_dispenser.latest_var().unwrap(),
+                                RangeConstraint::from_mask(1),
+                            );
+                        });
+                std::iter::once(constr).chain(extracted)
```

Comment:
So we'll end up with a constraint system with both constraints, right?

Reply:
I actually even tried adding `constr - extracted = 0` because that is additional knowledge that might be difficult to derive but it was not helpful for now.